### PR TITLE
Add initializer spec

### PR DESCRIPTION
**Pull request by 13013SwagR**

### CODE REVIEW

**Comment by yorugac** [comment]
Thanks, @13013SwagR :slightly_smiling_face: 
Initializer was deliberately set to use the runner's spec because they're both running k6, unlike the starter. I.e. one needs an image with k6 for both initializer and runner while starter doesn't depend on k6, at least at the moment. So initializer was set to use runner's spec in order to minimize entities and keep configurations simple. Normally, a user of the operator shouldn't even bother with initializer, unless there are errors in the script and initializer terminates with an error. I think if we are to change this and put initializer into a separate spec, we should be more clear on why we're doing that.

So one case is with dapr, described in #164, but I'm not fully clear dapr usage really depends on a separate spec for initializer or if it was resolved with container specification fix. I'd like to confirm this first. Sadly, I didn't have time to test this myself yet.

Are there other cases?

PS this discussion is more fit for an issue and not a PR I guess :see_no_evil:

**Comment by mcandeia** [comment]
> So one case is with dapr, described in #164, but I'm not fully clear dapr usage really depends on a separate spec for initializer or if it was resolved with container specification fix. I'd like to confirm this first. Sadly, I didn't have time to test this myself yet.

As I described there, "the problem" here is related on how kubernetes understand that a given job is completed. 

Since dapr is a sidecar (i.e one more container within the same pod) and the annotations for the runners and initializers are the same, the sidecar will be present in both, runners and initializers. The big difference is because when using runners I'm able to call the dapr shutdown [using its api](https://docs.dapr.io/operations/hosting/kubernetes/kubernetes-job/) ([see one example at dapr](https://github.com/dapr/dapr/blob/master/tests/perf/actor_id_scale/test.js#L57)) but as the initializers are totally out of my control I can't call shutdown, meaning that from the kubernetes point of view, the initializer job as a whole still pending even when the initializer container actually was successfully executed (because daprd still running). Furthermore, the operator leverages the kubernetes status to make sure the job is completed, see [here](https://github.com/grafana/k6-operator/blob/main/controllers/k6_initialize.go#L86).

Conclusion: The operator will not work when the initializer job has sidecar containers because the job isn't considered completed until all the containers are exited with code 0.

**Comment by yorugac** [comment]
Thank you for the clarifications, @mcandeia!

> Furthermore, the operator leverages the kubernetes status to make sure the job is completed, see [here](https://github.com/grafana/k6-operator/blob/main/controllers/k6_initialize.go#L86).

This part is unlikely to change even with resolution of #138. So yes, some alternative is clearly needed.

@13013SwagR,

>  I think the initializer Job should really be just an init container on the Starter Pod which solve my issue and the dapr one for the initializer.

This pod is required for the k6 Cloud output support. It can't be merged with the starter: these are clear separate stages that must be executed sequentially.

I've tested this with Envoy / scuttle: there's actually another bug which was also reported recently in #179 - I added the fix in PR #182.
To sum up, enabled Envoy will work with `disableLogging: true`. Otherwise, initializer output will be polluted with sidecar's logs and cannot be processed. This is "expected behaviour" at the moment. This is true regardless of initializer job being separate or not.

As for request to make initializer job separate, we have three cases so far:
1) dapr sidecar shutdown
2) disabling Istio for initializer only but keeping it for runners
3) downloading the script in init container of initializer (also connected to #118)

Given that, I think it makes sense to make initializer job separate at this point of time. More complex logic would be to make initializer job run only for specific cases but that requires some more thinking, esp. given the limitation with the logs (described above with Envoy).

File: pkg/resources/jobs/initializer.go [hunk 1]
```diff
@@ -27,28 +27,28 @@ func NewInitializerJob(k6 *v1alpha1.K6, argLine string) (*batchv1.Job, error) {
 		ports                        = append([]corev1.ContainerPort{{ContainerPort: 6565}}, k6.Spec.Ports...)
 	)
 
-	if k6.Spec.Runner.Image != "" {
-		image = k6.Spec.Runner.Image
+	if k6.Spec.Initializer.Image != "" {
+		image = k6.Spec.Initializer.Image
```
Comment (yorugac) [inline pkg/resources/jobs/initializer.go:31]:
```suggestion
	if k6.Spec.Initializer == nil {
		k6.Spec.Initializer = k6.Spec.Runner.DeepCopy()
	}

	if k6.Spec.Initializer.Image != "" {
		image = k6.Spec.Initializer.Image
```

File: api/v1alpha1/k6_types.go [hunk 2]
```diff
@@ -60,6 +60,7 @@ type K6Spec struct {
 	Separate    bool                   `json:"separate,omitempty"`
 	Arguments   string                 `json:"arguments,omitempty"`
 	Ports       []corev1.ContainerPort `json:"ports,omitempty"`
+	Initializer Pod                    `json:"initializer,omitempty"`
```
Comment (yorugac) [inline api/v1alpha1/k6_types.go:63]:
```suggestion
	Initializer *Pod                    `json:"initializer,omitempty"`
```

**Comment by yorugac** [comment]
Hello @13013SwagR, This issue with initializer job got a higher internal priority lately. No pressure but could you please let me know if you'll be able to work on this PR in the coming weeks? Thank you!

**Comment by 13013SwagR** [comment]
Hey sorry for the late response, will update in the next days.

**Comment by 13013SwagR** [comment]
Hey @yorugac, 
Did the rebase and implemented your suggestions
Let me know if there's anything else
Thanks

**Review by yorugac — APPROVED** [review]
Thank you @13013SwagR, for coming back to this!

It looks OK to me; will be merging in :+1:
